### ISSUE TITLE
Local Server bug with HTTP/2 Push

### ISSUE THREAD

**Issue by smnandre** [symfony-cli/symfony-cli#387]
TL;DR; as many people have experienced recently, the symfony local server seem to have trouble handling Link headers
* it tries to send them **_all_** via Push (even the rel=author or rel=next )
* the connection is never closed (it looks like it)

It works on Chrome/Safari only because they reject the push resources.

## (very simple) Tests

Two files: 
- index.php  (see per test)
- foo.css (see below)

<details>
	<summary>foo.css</summary>
```
@charset "utf-8";
h1 {
	color: green;
}
```
</details>

For every test:  symfony local server fully stopped / restarted

## Link preload

<details>
<summary>index.php</summary>

```
<?php 
header('Link: </foo.css>; rel=preload; as=style', false);
?>
<!doctype html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<link rel="stylesheet" href="/foo.css" type="text/css" />
		<title>Test</title>
	</head>
	<body>
		<h1>/foo.css</h1>
	</body>
</html>	

```
</details>

#### Firefox

Page + refresh (in vain)

```
[Web Server ] Dec  1 22:28:15 |DEBUG  | PHP    Reloading PHP versions
[Web Server ] Dec  1 22:28:15 |DEBUG  | PHP    Using PHP version 8.2.12 (from default version in $PATH)
[Web Server ] Dec  1 22:28:15 |INFO   | PHP    listening path="/opt/homebrew/Cellar/php/8.2.12_1/sbin/php-fpm" php="8.2.12" port=57182
[PHP-FPM    ] Dec  1 22:28:15 |NOTICE | FPM    fpm is running, pid 18852
[PHP-FPM    ] Dec  1 22:28:15 |NOTICE | FPM    ready to handle connections
[Web Server ] Dec  1 22:28:33 |INFO   | SERVER GET  (200) / ip="127.0.0.1" preloaded_resources=["/foo.css"]
[Web Server ] Dec  1 22:28:33 |INFO   | SERVER GET  (200) /foo.css
```

Server say the resource was preloaded but it was not, and then it's 100% stuck. Nothing to do except closing the browser or.... shuttind down the local server.

No response gets back.

<img width="531" alt="firefox requests" src="https://github.com/symfony-cli/symfony-cli/assets/1359581/b94b9bac-3769-4f46-bea3-25a642c28df0">

In the Firefox console, there is sometimes a quick warning about a missing frame ([related](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.6)?) , and an error.

Error

<img width="1244" alt="firefox error message" src="https://github.com/symfony-cli/symfony-cli/assets/1359581/b4e7eb1b-ccbb-4ba2-bf77-1e014cd980e9">

```
NS_ERROR_NOT_AVAILABLE: Component returned failure code: 0x80040111 (NS_ERROR_NOT_AVAILABLE) [nsIHttpChannelInternal.remoteAddress]
```

This comes from this code, and is related to remoteAddress & remotePort

<img width="921" alt="firefox error origin" src="https://github.com/symfony-cli/symfony-cli/assets/1359581/b242ed29-f282-4f50-b675-503c8cfbb110">

#### Chrome

Page + refresh

```
[Web Server ] Dec  1 22:26:52 |DEBUG  | PHP    Reloading PHP versions
[Web Server ] Dec  1 22:26:52 |DEBUG  | PHP    Using PHP version 8.2.12 (from default version in $PATH)
[Web Server ] Dec  1 22:26:52 |INFO   | PHP    listening path="/opt/homebrew/Cellar/php/8.2.12_1/sbin/php-fpm" php="8.2.12" port=56840
[PHP-FPM    ] Dec  1 22:26:52 |NOTICE | FPM    fpm is running, pid 18824
[PHP-FPM    ] Dec  1 22:26:52 |NOTICE | FPM    ready to handle connections
[Web Server ] Dec  1 22:26:59 |ERROR  | SERVER unable to preload links: feature not supported
[Web Server ] Dec  1 22:26:59 |INFO   | SERVER GET  (200) / ip="127.0.0.1"
[Web Server ] Dec  1 22:26:59 |INFO   | SERVER GET  (200) /foo.css
[Web Server ] Dec  1 22:26:59 |INFO   | SERVER GET  (200) /favicon.ico
[Web Server ] Dec  1 22:27:08 |ERROR  | SERVER unable to preload links: feature not supported
[Web Server ] Dec  1 22:27:08 |INFO   | SERVER GET  (200) / ip="127.0.0.1"
[Web Server ] Dec  1 22:27:08 |INFO   | SERVER GET  (200) /foo.css
[Web Server ] Dec  1 22:27:08 |INFO   | SERVER GET  (200) /favicon.ico

```

Everything works, and the stylesheet is fetched by the browser with high priority

## Pushing every link 

No browser naturally tries to fetch a Link without a preload or prefetch rel.

The symfony-cli currently pushes every ressource it finds (https://github.com/symfony-cli/symfony-cli/blob/1339531490a01e88cd10dac192b120ef1cf6e85d/local/http/push.go#L55), with no consideration for the "rel" attribute.

I'm not sure this is the main source of the problem, but it's definitively something that should not happen.

### Link rel=next (/foo.css)

<details>
<summary>index.php</summary>

```
<?php 
header('Link: </foo.css>; rel=next', false);
?>
<!doctype html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<link rel="stylesheet" href="/foo.css" type="text/css" />
		<title>Test</title>
	</head>
	<body>
		<h1>/foo.css</h1>
	</body>
</html>	
```
</details>

#### Firefox

```
[Web Server ] Dec  1 22:17:23 |DEBUG  | PHP    Reloading PHP versions
[Web Server ] Dec  1 22:17:23 |DEBUG  | PHP    Using PHP version 8.2.12 (from default version in $PATH)
[PHP-FPM    ] Dec  1 22:17:23 |NOTICE | FPM    fpm is running, pid 18661
[PHP-FPM    ] Dec  1 22:17:23 |NOTICE | FPM    ready to handle connections
[Web Server ] Dec  1 22:17:23 |INFO   | PHP    listening path="/opt/homebrew/Cellar/php/8.2.12_1/sbin/php-fpm" php="8.2.12" port=54493
[Web Server ] Dec  1 22:17:38 |INFO   | SERVER GET  (200) / ip="127.0.0.1" preloaded_resources=["/foo.css"]
[Web Server ] Dec  1 22:17:38 |INFO   | SERVER GET  (200) /foo.css
```

Then Firefox is stuck, no HTTP request is handled after this one (neither the fake /foo.css or the /favicon.ico), and the tab is frozen.

#### Chrome

```
[Web Server ] Dec  1 22:20:30 |ERROR  | SERVER unable to preload links: feature not supported
[Web Server ] Dec  1 22:20:30 |INFO   | SERVER GET  (200) / ip="127.0.0.1"
[Web Server ] Dec  1 22:20:30 |INFO   | SERVER GET  (200) /foo.css
[Web Server ] Dec  1 22:20:30 |INFO   | SERVER GET  (200) /favicon.ico
```

As Chrome refuses the resource pushed by the server, it naturally find the CSS in the HTML and download it. No problem here.

### Link rel=next  (missing file)

<details>
<summary>index.php</summary>

```
<?php 
header('Link: </babar.css>; rel=next', false);
?>
<!doctype html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<link rel="stylesheet" href="/foo.css" type="text/css" />
		<title>Test</title>
	</head>
	<body>
		<h1>/foo.css</h1>
	</body>
</html>	
```
</details>

#### Firefox

```
[Web Server ] Dec  1 22:07:57 |DEBUG  | PHP    Reloading PHP versions
[Web Server ] Dec  1 22:07:58 |DEBUG  | PHP    Using PHP version 8.2.12 (from default version in $PATH)
[Web Server ] Dec  1 22:07:58 |INFO   | PHP    listening path="/opt/homebrew/Cellar/php/8.2.12_1/sbin/php-fpm" php="8.2.12" port=52160
[PHP-FPM    ] Dec  1 22:07:58 |NOTICE | FPM    fpm is running, pid 18487
[PHP-FPM    ] Dec  1 22:07:58 |NOTICE | FPM    ready to handle connections
[Web Server ] Dec  1 22:08:13 |INFO   | SERVER GET  (200) / ip="127.0.0.1" preloaded_resources=["/babar.css"]
[Web Server ] Dec  1 22:08:13 |ERROR  | SERVER unable to preload links: http2: recursive push not allowed
[Web Server ] Dec  1 22:08:13 |INFO   | SERVER GET  (200) /babar.css ip="127.0.0.1"
```

The server pretend it preloaded /babar.css ... but the file does not exists. 

And then Firefox still tries to GET that file

#### Safari

```
[Web Server ] Dec  1 22:11:16 |DEBUG  | PHP    Reloading PHP versions
[Web Server ] Dec  1 22:11:16 |DEBUG  | PHP    Using PHP version 8.2.12 (from default version in $PATH)
[Web Server ] Dec  1 22:11:16 |INFO   | PHP    listening path="/opt/homebrew/Cellar/php/8.2.12_1/sbin/php-fpm" php="8.2.12" port=52995
[PHP-FPM    ] Dec  1 22:11:16 |NOTICE | FPM    fpm is running, pid 18621
[PHP-FPM    ] Dec  1 22:11:16 |NOTICE | FPM    ready to handle connections
[Web Server ] Dec  1 22:11:23 |ERROR  | SERVER unable to preload links: feature not supported
[Web Server ] Dec  1 22:11:23 |INFO   | SERVER GET  (200) / ip="127.0.0.1"
[Web Server ] Dec  1 22:11:23 |INFO   | SERVER GET  (200) /foo.css
```

Safari, as Chrome before, refuses the Push and then do its job naturally.

## No problem with Firefox when..

Note: to be sure that what not something only caused by Firefox, i put all those tests on a live server, and Firefox has no problem.

I also tried some tests in an old Docker project using hosts mapped in my /etc/hosts + Traeffic as a reverse proxy.. no problem either.

So it is related to the localhost, the symfony server, the certificates, ... but there is a bug somewhere here :)

Please tell me if i can do anything

**Comment by smnandre** [comment]
Sadly no.... or not for the main problem..

With #392 the rel=next are not pushed anymore (thanks for that).

But the problem with rel=preload remains (logically)

I'd suggest we do not push any resource, as Firefox is the only browser accepting them, and probably with such requirements i'm not sure it's worth the pain for a local server. 


To recap, with an existing file foo.css at the root of the project, relative link in header does not work (absolute one does).


`Link: <https://127.0.0.1:8007/foo.css>; rel=preload`
<details>
<summary>
Does work: 
</summary>

```
<?php
header('Link: <https://127.0.0.1:8007/foo.css>; rel=preload; as=style', false);
?>
<!doctype html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<link rel="stylesheet" href="/foo.css" type="text/css" />
		<title>Test</title>
	</head>
	<body>
		<h1>/foo.css</h1>
	</body>
</html>
```

</details>


`Link: </foo.css>; rel=preload`
<details>
<summary>Does not work: </summary>

```php
<?php
header('Link: </foo.css>; rel=preload; as=style', false);
?>
<!doctype html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<link rel="stylesheet" href="/foo.css" type="text/css" />
		<title>Test</title>
	</head>
	<body>
		<h1>/foo.css</h1>
	</body>
</html>
```

</details>

**Comment by smnandre** [comment]
@tucksaun 

I have missed some releases since then i'm sorry. 

I just tried the fresh 5.8.3 ... and it's perfect now ! 

And is it me or the PHP has gained a lot of performance / efficience ? I see local requests (via proxy / tld / etc) in less than 15ms oO.

Thanks a lot for all that :)
